sql/src/main/java/io/crate/planner/node/management/ExplainPlan.java

```diff
@@ -167,18 +173,83 @@ private TransportCollectProfileOperation getRemoteCollectOperation(DependencyCar
-        nodeTimings.forEach(executionTimingsMap::put);
+        executionTimingsMap.put("Phases", phasesTimings);
+        resultNodeTimings.forEach(executionTimingsMap::put);
         executionTimingsMap.put("Total", apeTimings.get(Phase.Execute.name()));
 
         mapBuilder.put(Phase.Execute.name(), executionTimingsMap.immutableMap());
         return new Row1(mapBuilder.immutableMap());
     }
 
+    private static Map<String, Object> extractPhasesTimingsFrom(Map<String, Map<String, Object>> nodesTimings,
+                                                                NodeOperationTree operationTree) {
+        Map<String, Object> allPhases = new TreeMap<>();
+        for (NodeOperation operation : operationTree.nodeOperations()) {
+            ExecutionPhase phase = operation.executionPhase();
+            String phaseName = getPhaseName(phase);
+            Map<String, Object> phaseTimingsAcrossNodes = getPhaseTimingsAcrossNodes(phaseName, nodesTimings);
+
+            if (!phaseTimingsAcrossNodes.isEmpty()) {
+                allPhases.put(phaseName, MapBuilder.newMapBuilder().put("nodes", phaseTimingsAcrossNodes).immutableMap());
+            }
+        }
+
+        ExecutionPhase leafExecutionPhase = operationTree.leaf();
+        String phaseName = getPhaseName(leafExecutionPhase);
+        Map<String, Object> phaseTimingsAcrossNodes = getPhaseTimingsAcrossNodes(phaseName, nodesTimings);
+
+        if (!phaseTimingsAcrossNodes.isEmpty()) {
+            allPhases.put(phaseName, MapBuilder.newMapBuilder().put("nodes", phaseTimingsAcrossNodes).immutableMap());
+        }
+
+        return allPhases;
+    }
+
+    private static String getPhaseName(ExecutionPhase phase) {
+        return String.format(Locale.ROOT, "%d-%s", phase.phaseId(), phase.name());
+    }
+
+    private static Map<String, Object> getPhaseTimingsAcrossNodes(String phaseName,
+                                                                  Map<String, Map<String, Object>> nodesTimings) {
+        MapBuilder<String, Object> timingsForPhaseAcrossNodes = MapBuilder.newMapBuilder();
+        for (String operationNode : nodesTimings.keySet()) {
+            Map<String, Object> timingsForNode = nodesTimings.get(operationNode);
+            if (timingsForNode != null) {
+                Object phaseTiming = timingsForNode.get(phaseName);
+                if (phaseTiming != null) {
+                    timingsForPhaseAcrossNodes.put(operationNode, phaseTiming);
+                }
+            }
+        }
+
+        return timingsForPhaseAcrossNodes.immutableMap();
+    }
+
+    private static Map<String, Map<String, Object>> getNodeTimingsWithoutPhases(Set<String> phasesNames,
+                                                                                Map<String, Map<String, Object>> nodeTimings) {
+        Map<String, Map<String, Object>> nodeTimingsWithoutPhases = new HashMap<>(nodeTimings.size());
```

Comment:
Could we use the same MapBuilder pattern as you used in the `getPhaseTimingsAcrossNodes` method?

Reply:
I thought it's a bit wasteful to create the additional Builder objects since we need to also copy the inside Map (so it'd be 2 extra MapBuilder objects).
Do you think it's worth it?